src/main/java/net/sf/mzmine/util/scans/ScanUtils.java

```diff
@@ -336,6 +317,69 @@ public static double calculateTIC(Scan scan, Range<Double> mzRange) {
 
   }
 
+  /**
+   * sort the data points by their m/z value. This method should be called before using other search methods to do binary search in
+   * logarithmic time.
+   * @param dataPoints spectrum that should be sorted
+   */
+  public static void sortDataPointsByMz(DataPoint[] dataPoints) {
+    Arrays.sort(dataPoints, Comparator.comparingDouble(DataPoint::getMZ));
+  }
+
+  /**
+   * Returns the index of the datapoint with lowest m/z within the given datapoints which is within the given mass range
+   * @param dataPoints sorted(!) list of datapoints
+   * @param mzRange m/z range to search in
+   * @return index of datapoint or -1, if no datapoint is in range
+   */
+  public static int findFirstPeakWithin(DataPoint[] dataPoints, Range<Double> mzRange) {
+    final int insertionPoint = Arrays.binarySearch(dataPoints, new SimpleDataPoint(mzRange.lowerEndpoint(), 0d), (u,v)->Double.compare(u.getMZ(),v.getMZ()));
+    if (insertionPoint<0) {
+      final int k = -insertionPoint - 1;
+      if (k < dataPoints.length && mzRange.contains(dataPoints[k].getMZ())) return k;
+      else return -1;
+    } else {
+      return insertionPoint;
+    }
+  }
+
+  /**
+   * Returns the index of the datapoint with largest m/z within the given datapoints which is within the given mass range
+   * @param dataPoints sorted(!) list of datapoints
+   * @param mzRange m/z range to search in
+   * @return index of datapoint or -1, if no datapoint is in range
+   */
+  public static int findLastPeakWithin(DataPoint[] dataPoints, Range<Double> mzRange) {
+    final int insertionPoint = Arrays.binarySearch(dataPoints, new SimpleDataPoint(mzRange.upperEndpoint(), 0d), (u,v)->Double.compare(u.getMZ(),v.getMZ()));
+    if (insertionPoint<0) {
+      final int k = -insertionPoint - 2;
+      if (k >= 0 && mzRange.contains(dataPoints[k].getMZ())) return k;
+      else return -1;
+    } else {
+      return insertionPoint;
+    }
+  }
+
+  /**
+   * Returns the index of the datapoint with highest intensity within the given datapoints which is within the given mass range
+   * @param dataPoints sorted(!) list of datapoints
+   * @param mzRange m/z range to search in
+   * @return index of datapoint or -1, if no datapoint is in range
+   */
+  public static int findMostIntensivePeakWithin(DataPoint[] dataPoints, Range<Double> mzRange) {
```

Comment:
A minor thing, but the correct term is "intense", not "intensive"

Reply:
thx